sqlflow_models/gcn.py

```diff
@@ -0,0 +1,369 @@
+                        indices=np.array(list(zip(adjacency.row, adjacency.col)), dtype=np.int64),
+                        values=tf.cast(adjacency.data, tf.float32),
+                        dense_shape=adjacency.shape)
+        adjacency = tf.sparse.reorder(adjacency)
+        
+        total_num = features.shape[0]
+        train_num = round(total_num*self.train_ratio)
+        eval_num = round(total_num*self.eval_ratio)
+        train_index = np.arange(train_num)
+        val_index = np.arange(train_num, train_num+eval_num)
+        test_index = np.arange(train_num+eval_num, total_num)
+
+        self.train_mask = np.zeros(total_num, dtype = np.bool)
+        self.val_mask = np.zeros(total_num, dtype = np.bool)
+        self.test_mask = np.zeros(total_num, dtype = np.bool)
+        self.train_mask[train_index] = True
+        self.val_mask[val_index] = True
+        self.test_mask[test_index] = True
+
+        print('Dataset has {} nodes, {} edges, {} features.'.format(features.shape[0], edges.shape[0], features.shape[1]))
+
+        return features, labels, adjacency, nf_shape, na_shape
+    
+    def loss_func(self, model, x, y, train_mask, training=True):
+        '''Customed loss function for the model.'''
+
+        y_ = model(x, training=training)
+
+        test_mask_logits = tf.gather_nd(y_, tf.where(train_mask))
+        masked_labels = tf.gather_nd(y, tf.where(train_mask))
+
+        return loss(labels=masked_labels, output=test_mask_logits)
+
+    def grad(self, model, inputs, targets, train_mask):
+        '''Calculate the gradients of the parameters.'''
+        with tf.GradientTape() as tape:
+            loss_value = self.loss_func(model, inputs, targets, train_mask)
+        
+        return loss_value, tape.gradient(loss_value, model.trainable_variables)
+    
+    def test(self, mask, return_loss=False):
+        '''Test the results on the model. Return accuracy'''
+        logits = self.predict(data=[self.features, self.adjacency])
+
+        test_mask_logits = tf.gather_nd(logits, tf.where(mask))
+        masked_labels = tf.gather_nd(self.labels, tf.where(mask))
+
+        ll = tf.equal(tf.argmax(masked_labels, -1), tf.argmax(test_mask_logits, -1))
+        accuarcy = tf.reduce_mean(tf.cast(ll, dtype=tf.float32))
+
+        if return_loss:
+            loss_value = loss(labels=masked_labels, output=test_mask_logits)
+            return [loss_value, accuarcy]
+
+        return accuarcy
+
+    def sqlflow_train_loop(self, x):
+        """Customized training function."""
+        # load data
+        ids, ids_check, features, labels, edges, edge_check = list(), dict(), list(), list(), list(), dict()
+        from_node = 0
+        for inputs, label in x:
+            id = inputs['id'].numpy().astype(np.int32)
+            feature = inputs['features'].numpy().astype(np.float32)
+            from_node = inputs['from_node_id'].numpy().astype(np.int32)
+            to_node = inputs['to_node_id'].numpy().astype(np.int32)
```

Comment:
Yeah sure, this will be more flexible when people create dataset with different column names. I will update this later :)

Reply:
One thing I have to mention is that, `id`, `from_node_id` and `to_node_id` should be `integer` since they are treated as the index of the node in a certain order. Otherwise, it may cause some problems when we build the graph with this code. 

I will make this clear in the GCN documentation later.